batchflow/models/tf/layers/core.py

```diff
@@ -50,25 +51,121 @@ def __call__(self, inputs):
+            if self.op in ['avg', 'average', 'mean']:
+                return tf.reduce_mean(tf.stack(inputs, axis=0), axis=0, name='combine-mean')
+            if self.op in ['sum', 'add']:
+                return tf.add_n(inputs, name='combine-sum')
+            if self.op in ['softsum', 'convsum']:
+                filters = inputs[0].get_shape().as_list()[axis]
+                for i in range(1, len(inputs)):
+                    inputs[i] = Conv(filters=filters, kernel_size=1, name='combine-conv', **self.kwargs)(inputs[i])
+                return tf.add_n(inputs, name='combine-softsum')
+            raise ValueError('Unknown operation {}.'.format(self.op))
+
+
+
+class BaseDropout(Layer):
+    """ Base class for dropout layers.
+
+    Parameters
+    ----------
+    dropout_rate : float, tf.Tensor, callable
+        If float or Tensor, then fraction of the input units to drop.
+        If callable, then function to be called on `global_step`. Must return tensor of size 1.
+
+    multisample: bool, number, sequence, tf.Tensor
+        If evaluates to True, then either multiple dropout applied to the whole batch and then averaged, or
+        batch is split into multiple parts, each passed through dropout and then concatenated back.
+
+        If True, then two different dropouts are applied to whole batch.
+        If integer, then that number of different dropouts are applied to whole batch.
+        If float, then batch is split into parts of `multisample` and `1 - multisample` sizes.
+        If sequence of ints, then batch is split into parts of given sizes. Must sum up to the batch size.
+        If sequence of floats, then each float means proportion of sizes in batch and must sum up to 1.
+        If Tensor, then it is used as the second parameter for splitting function, see
+        `tf.split <https://www.tensorflow.org/api_docs/python/tf/split>`_,.
+    """
+    def __init__(self, dropout_rate, multisample=False, global_step=None, **kwargs):
         self.dropout_rate = dropout_rate
+        self.global_step = global_step
+        self.multisample = multisample
         self.kwargs = kwargs
 
     def __call__(self, inputs, training):
-        return K.Dropout(rate=self.dropout_rate, **self.kwargs)(inputs, training)
+        if callable(self.dropout_rate):
+            step = tf.cast(self.global_step, dtype=tf.float32)
+            self.dropout_rate = self.dropout_rate(step)
+        d_layer = self.LAYER(rate=self.dropout_rate, **self.kwargs)
+
+        if self.multisample is not False:
+            if self.multisample is True:
+                self.multisample = 2
+            elif isinstance(self.multisample, float):
+                self.multisample = [self.multisample, 1 - self.multisample]
+
+            if isinstance(self.multisample, int): # dropout to the whole batch, then average
+                dropped = [d_layer(inputs, training) for _ in range(self.multisample)]
+                output = Combine(op='avg')(dropped)
```

Comment:
I think we need to add an option to apply activation to each sample before combine operation.

Reply:
If one needs an activation before combine, one can do it before calling combine.
To include activation in combine, one can add necessary args (layout, activation, etc)